### PR TITLE
planner, table: list partition support prune non point range

### LINKED ISSUE #56673: List partition can't prune access partition with greater or less condition
## Bug Report

Please answer these questions before submitting your issue. Thanks!

### 1. Minimal reproduce step (Required)

```sql
CREATE TABLE `t` (
  `a` int NOT NULL,
  `b` varchar(255) DEFAULT NULL,
  PRIMARY KEY (`a`)
) ENGINE=InnoDB DEFAULT CHARSET=utf8mb4 COLLATE=utf8mb4_0900_ai_ci
/*!50100 PARTITION BY LIST (`a`)
(PARTITION p0 VALUES IN (null, 1,2,3) ENGINE = InnoDB,
 PARTITION p2 VALUES IN (-1,-2,-3) ENGINE = InnoDB,
 PARTITION p1 VALUES IN (100,101,102) ENGINE = InnoDB) */;
explain select * from t where a > 0;  -- access p0,p1
explain select * from t where a < 0;  -- access p2
```

### 2. What did you expect to see? (Required)

prune success

```
mysql> explain select * from t where a > 0;
+----+-------------+-------+------------+-------+---------------+---------+---------+------+------+----------+-------------+
| id | select_type | table | partitions | type  | possible_keys | key     | key_len | ref  | rows | filtered | Extra       |
+----+-------------+-------+------------+-------+---------------+---------+---------+------+------+----------+-------------+
|  1 | SIMPLE      | t     | p0,p1      | range | PRIMARY       | PRIMARY | 4       | NULL |    1 |   100.00 | Using where |
+----+-------------+-------+------------+-------+---------------+---------+---------+------+------+----------+-------------+
1 row in set, 1 warning (0.00 sec)
```

### 3. What did you see instead (Required)

access all partitions

```
mysql> explain select * from t where a > 0;
+------------------------+---------+-----------+---------------+------------------------------------------------+
| id                     | estRows | task      | access object | operator info                                  |
+------------------------+---------+-----------+---------------+------------------------------------------------+
| TableReader_6          | 3333.33 | root      | partition:all | data:TableRangeScan_5                          |
| └─TableRangeScan_5     | 3333.33 | cop[tikv] | table:t       | range:(0,+inf], keep order:false, stats:pseudo |
+------------------------+---------+-----------+---------------+------------------------------------------------+
2 rows in set (0.01 sec)
```

### 4. What is your TiDB version? (Required)

```
mysql> select tidb_version();
+--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
| tidb_version()                                                                                                                                                                                                                                                                 |
+--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
| Release Version: v8.4.0-alpha-419-g3df0f2e927-dirty
Edition: Community
Git Commit Hash: 3df0f2e927a4d1d5b9431ab4dc0f5a2a71f04786
Git Branch: master
UTC Build Time: 2024-10-16 02:36:14
GoVersion: go1.23.1
Race Enabled: false
Check Table Before Drop: false
Store: unistore |
+--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
1 row in set (0.01 sec)
```

### PR DESCRIPTION
**Pull request by Defined2014**
### What problem does this PR solve?

Issue Number: close #56673

Problem Summary:

### Check List

Tests 

- [x] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No need to test
  > - [ ] I checked and no code files have been changed.
  > 

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
None
```

### CODE REVIEW

File: pkg/planner/core/casetest/partition/testdata/integration_partition_suite_out.json [hunk 1]
```diff
@@ -91,24 +91,14 @@
       {
         "SQL": "explain format = 'brief' select * from tlist where a not in (0, 1, 2, 3, 4, 5, 6, 7, 8)",
         "DynamicPlan": [
-          "TableReader 3583.33 root partition:all data:Selection",
+          "TableReader 3583.33 root partition:p3 data:Selection",
```
Comment (mjonss) [inline pkg/planner/core/casetest/partition/testdata/integration_partition_suite_out.json:94]: Does the range optimizer check the list and returns two ranges (<0 and >8), which results in the new pruning result?

File: pkg/table/tables/partition.go [hunk 2]
```diff
@@ -960,7 +966,7 @@ func (lp *ForListPruning) buildListColumnsPruner(ctx expression.BuildContext,
 // colIdx is the column index in the list columns.
 func (lp *ForListPruning) buildListPartitionValueMap(ctx expression.BuildContext, defs []model.PartitionDefinition,
 	schema *expression.Schema, names types.NameSlice, p *parser.Parser) error {
-	lp.valueMap = map[int64]int{}
+	lp.valueMap = btree.NewG[*btreeListItem](btreeDegree, func(a, b *btreeListItem) bool { return a.key < b.key })
```
Comment (mjonss) [inline pkg/table/tables/partition.go:969]: Are there any tests for `bigint unsigned` List partitioning columns? And how `bigint signed` handles comparison between MaxInt64 and MaxUInt64?

**Review by mjonss — APPROVED** [review]
LGTM, just a curious question about the range optimizer as well as how UNSIGNED bigint partitioning column is handled.

File: pkg/planner/core/casetest/partition/testdata/integration_partition_suite_out.json (hunk 1 above)
Comment (Defined2014) [reply pkg/planner/core/casetest/partition/testdata/integration_partition_suite_out.json:94]: Yes.

File: pkg/table/tables/partition.go (hunk 2 above)
Comment (Defined2014) [reply pkg/table/tables/partition.go:969]: Good point, I will fix it and add some test cases.

File: pkg/table/tables/partition.go [hunk 3]
```diff
@@ -425,7 +426,7 @@ type ForListPruning struct {
 	// PruneExprCols is the columns of PruneExpr, it has removed the duplicate columns.
 	PruneExprCols []*expression.Column
 	// valueMap is column value -> partition idx, uses to locate list partition.
-	valueMap map[int64]int
+	valueMap *btree.BTreeG[*btreeListItem]
```
[the remark this answers is not in the document]
Comment (Defined2014) [reply pkg/table/tables/partition.go]: Done

**Review by winoros — APPROVED** [review]

**Review by tangenta — APPROVED** [review]

**Pull request merged**